universe/Condition.cpp

```diff
@@ -9956,6 +9965,249 @@ unsigned int Not::GetCheckSum() const {
+            if (!non_matches.empty()) {
+                // i.e. the current operand should be used negated to move
+                found_sth = true;
+            }
+            // recreate previous state by moving back temp_non matches to non_matches
+            FCMoveContent(temp_non_matches, non_matches);
+
+            if (!found_sth) {
+                 // Also have to look at the other local candidates if there was a match
+                ObjectSet temp_matches;
+                temp_matches.reserve(matches.size());
+                operand->Eval(local_context, temp_matches, matches, NON_MATCHES);
+                if (!temp_matches.empty()) {
+                    // i.e. the current operand should be used negated to move
+                    found_sth = true;
+                }
+                // recreate previous state
+                FCMoveContent(temp_matches, matches);
+            }
+
+            if (found_sth) {
+                // descent into subcondition for NON_MATCHES
+                operand->Eval(local_context, matches, non_matches, NON_MATCHES);
+                break;
+            }
+            // try the next operand
+        }
+        // If NON_MATCHES could not find any matches it should match all
+        if (!found_sth)
+            FCMoveContent(matches, non_matches);
+    } else /*(search_domain == MATCHES)*/ {
+        // check all operand conditions on all objects in the matches set, collecting non matches temporarily
+        // stop if the operand condition matches at least one item
+        // if the operand condition does not match, restore the matches set by moving the temporary non-matches back
+        ObjectSet temp_non_matches;
+        temp_non_matches.reserve(matches.size());
+        // try matching operand conditions until one mateches
+        // use a fresh set of possible matches
+        for (auto& operand : m_operands) {
+            // move back temp_non_matches
+            FCMoveContent(temp_non_matches, matches);
+            // move non matches from_matches to temp_non_matches
+            operand->Eval(local_context, matches, temp_non_matches, MATCHES);
+            if (!matches.empty()) break;
+            // Check if other LocalCandidates (in the non_matches set) match
+            operand->Eval(local_context, matches, non_matches, NON_MATCHES);
+            if (!matches.empty()) {
+                // This operand should return the result, but no matches exist
+                // So all objects should go into the non_matches set
+                // move all matches to non_matches here
+                FCMoveContent(matches, non_matches);
+                // stop finding the operand
+                // temp_non_matches must still be moved to non_matches
+                break;
+            }
+        }
+
+        // non matching items were already moved from matches to temp_non_matches
+        // move non matching items from temp_non_matches to non_matches
```

Comment:
// Objects in the input matches set that did not match the chosen operand condition were moved above into temp_non_matches. Move them into non_matches (since they aren't matched by this condition).